src/from_exception.cpp

```diff
@@ -4,6 +4,147 @@
 // accompanying file LICENSE_1_0.txt or copy at
 // http://www.boost.org/LICENSE_1_0.txt)
 
+#if defined(_MSC_VER)
+
+#include <boost/stacktrace/safe_dump_to.hpp>
+#include <windows.h>
+
+extern "C" void** __cdecl __current_exception(); // exported from vcruntime.dll
+#define _pCurrentException static_cast<PEXCEPTION_RECORD>(*__current_exception())
+
+namespace {
+
+constexpr std::size_t kStacktraceDumpSize = 4096;
+
+struct thrown_info {
+  ULONG_PTR object;
+  char* dump;
+};
+
+struct exception_data {
+  bool capture_stacktraces_at_throw = true;
+  unsigned count = 0;
+  thrown_info* info = nullptr;
+
+  ~exception_data() {
+    HANDLE hHeap = GetProcessHeap();
+    for (unsigned i = 0; i < count; ++i) {
+      HeapFree(hHeap, 0, info[i].dump);
+    }
+    HeapFree(hHeap, 0, info);
+  }
+};
+
+thread_local exception_data data;
+
+inline bool PER_IS_MSVC_EH(PEXCEPTION_RECORD p) {
+  DWORD EH_EXCEPTION_NUMBER = 0xE06D7363;
+  ULONG_PTR EH_MAGIC_NUMBER1 = 0x19930520;
+
+  return p->ExceptionCode == EH_EXCEPTION_NUMBER &&
+    (p->NumberParameters == 3 || p->NumberParameters == 4) &&
+    p->ExceptionInformation[0] == EH_MAGIC_NUMBER1;
+}
+
+inline ULONG_PTR PER_PEXCEPTOBJ(PEXCEPTION_RECORD p) {
```

Comment:
use lowercase to avoid collisions with macro

Reply:
It is taken from `ehdata.h` in vcruntime (which is a macro) with slight change. Since we don't include the header directly, I would like to keep as it is. It is the same case as `PER_IS_MSVC_EH`.